cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizerService.java

```diff
@@ -59,6 +64,92 @@
       .setHttpHandlers(new ArtifactLocalizerHttpHandlerInternal(cConf));
 
     httpService = builder.build();
+    cacheDir = Paths.get(cConf.get(CFG_LOCAL_DATA_DIR));
+    cacheCleanupInterval = cConf.getInt(Constants.ArtifactLocalizer.CACHE_CLEANUP_INTERVAL_MIN);
+  }
+
+  @Override
+  protected void runOneIteration() throws Exception {
+    cleanupArtifactCache(cacheDir.resolve("artifacts").toFile());
+  }
+
+  /**
+   * Recursively looks for out of date artifact jar files in cacheDir. Once a jar file is found, we will first attempt
+   * to delete the unpacked directory for this jar (if it exists) and then delete the jar if and only if that was
+   * successful.
+   *
+   * @param cacheDir The directory that contains the cached jar files
+   */
+  private void cleanupArtifactCache(@NotNull File cacheDir) {
+    List<Long> timestamps = new ArrayList<>();
+
+    File[] files = cacheDir.listFiles();
+    if (files == null) {
+      return;
+    }
+
+    // Scan all files in the current directory, if its a directory recurse, otherwise add the timestamp value to a list
+    for (File file : files) {
+      if (file.isDirectory()) {
+        cleanupArtifactCache(file);
+        continue;
+      }
+
+      // We're only interested in jar files
+      if (!file.getName().endsWith(".jar")) {
+        continue;
+      }
+      try {
+        timestamps.add(Long.parseLong(FilenameUtils.removeExtension(file.getName())));
+      } catch (NumberFormatException e) {
+        // If we encounter a file that doesn't have a timestamp as the filename
+        LOG.warn("Encountered unexpected file during artifact cache cleanup: {}. This file will be deleted.",
+                 file.getPath());
+        file.delete();
+      }
+    }
+
+    // If there are less than two timestamp files then we don't need to clean anything
+    if (timestamps.size() < 2) {
```

Comment:
This 2 is used to skip processing when we have 0 (the code would fail when it tries to get the max) or 1 (the code would work its just a waste of time). The code that comes after this will always delete all file except the max timestamp.

I can refactor this to keep around a set number of artifacts but I dont see a scenario where we would want to hold on to an out of date jar. Do you have a use case in mind?